src/utils/programs.ts

```diff
@@ -0,0 +1,78 @@
+import type * as ts from "typescript";
+
+import fs from "node:fs";
+import path from "node:path";
+
+type TSModule = typeof ts;
+
+interface ReadConfigFile {
+	config: {
+		compilerOptions: {
+			module: string;
+		};
+	};
+}
+
+function createProgram(configFile: string, ts: TSModule): ts.Program {
+	const projectDirectory = path.dirname(configFile);
+	const { config } = ts.readConfigFile(configFile, (...args) => {
+		return ts.sys.readFile(...args);
+	}) as ReadConfigFile;
+	const parseConfigHost: ts.ParseConfigHost = {
+		fileExists: fs.existsSync,
+		// eslint-disable-next-line @typescript-eslint/unbound-method
+		readDirectory: ts.sys.readDirectory,
+		readFile: (file) => fs.readFileSync(file, "utf8"),
+		useCaseSensitiveFileNames: true,
+	};
+	const parsed = ts.parseJsonConfigFileContent(
+		config,
+		parseConfigHost,
+		path.resolve(projectDirectory),
+		{
+			noEmit: true,
+		},
+	);
+
+	// If the TypeScript version is too old to handle the "node16" module option,
+	// we can still run tests falling back to commonjs/node.
+	// Once 4.7 is out of DefinitelyTyped support (2024-05) we can remove this.
+	// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html
+	if (
+		config.compilerOptions.module.toLowerCase() === "node16" &&
+		parsed.options.module === undefined
+	) {
+		parsed.options.module = ts.ModuleKind.CommonJS;
+		// eslint-disable-next-line deprecation/deprecation
+		parsed.options.moduleResolution = ts.ModuleResolutionKind.NodeJs;
+	}
+
+	const host = ts.createCompilerHost(parsed.options, true);
+	return ts.createProgram(parsed.fileNames, parsed.options, host);
```

Comment:
I couldn't figure out how to get the programs created by this to read from the files set up by the rule tester. https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/actions/runs/7283460649/job/19847369501 > https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/pull/123#discussion_r1433402020:

> Right now in tests this is logging:
> 
> ```
> sourceFile.getText() // (stub file for RuleTester programs)
> ```
> 
> 🤔 I couldn't figure out how to get the program to read from virtual memory. Grr.

For now the `versions` option isn't unit tested... Not great.